### PR TITLE
Add device logic handling for functions which allow scalar inputs as tensors

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86149
* #86144

Some functions allow scalars as tensor inputs. Add handling for them in device logic.

Fix for https://github.com/pytorch/torchdynamo/issues/1445

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -567,6 +567,14 @@ def merge_devices(t):
         tree_map(merge_devices, args)
         tree_map(merge_devices, kwargs)
 
+        # some functions that allow Python numbers to bind to Tensors
+        # if we have failed to find a device, and we're running one of these operators,
+        # we must have scalar only inputs
+        if torch._C._should_allow_numbers_as_tensors(
+            func.name().split("::")[-1].split(".")[0]
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:574]: I feel that func already has an API for getting this directly, and if it does not, you should add one.

**Review by ezyang — APPROVED** [review]

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (eellison) [reply torch/_subclasses/fake_tensor.py]: `unqualified_name` I guess ?

**Review by bdhirsh — APPROVED** [review]

**Pull request closed without merging**